src/Diff/Types.hs

```diff
@@ -0,0 +1,199 @@
+
+{-# LANGUAGE DuplicateRecordFields      #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+
+module Diff.Types where
+
+import Data.List.NonEmpty (NonEmpty(..))
+import Data.Map (Map)
+import Data.Set (Set)
+import Data.Text (Text)
+import Nix.Derivation (DerivationOutput (..))
+
+import qualified Patience
+import Data.Aeson
+import GHC.Generics (Generic)
+import Data.Foldable
+import Data.Aeson.Types (listValue)
+
+data Changed a = Changed { before :: a, now :: a }
+
+instance ToJSON a => ToJSON (Changed a) where
+  toJSON = changedToJSON toJSON
+
+newtype TextDiff f = TextDiff {unTextDiff :: f (Patience.Item Text)}
+
+instance Foldable f => ToJSON (TextDiff f) where
+  toJSON = listValue toJSONItem . squashDiff . toList . unTextDiff
+    where
+      -- We want to have the longest difference spans,
+      -- instead of printing each word as one item
+      squashDiff (Patience.Old a : Patience.Old b : xs) =
+        squashDiff (Patience.Old (a <> b) : xs)
+      squashDiff (Patience.New a : Patience.New b : xs) =
+        squashDiff (Patience.New (a <> b) : xs)
+      squashDiff (Patience.Both a _ : Patience.Both b _ : xs) =
+        let ab = a <> b in squashDiff (Patience.Both ab ab : xs)
+      squashDiff (x: xs) = x : squashDiff xs
+      squashDiff [] = []
+
+      toJSONItem (Patience.Old a)    = object
+        [ "type" .= ("Old" :: Text), "content" .= a]
+      toJSONItem (Patience.New a)    = object
+        [ "type" .= ("New" :: Text), "content" .= a]
+      toJSONItem (Patience.Both a _) = object
+        [ "type" .= ("Both" :: Text), "content" .= a]
+
+-- Helpfull aliaces
+
+type OutputHash = Text
+
+type Platform = Text
+
+type Builder = Text
+
+type Argument = Text
+
+-- Derivation diff
+
+data DerivationDiff
+  = DerivationsAreTheSame
+  | AlreadyCompared
+  | NamesDontMatch   { outputStructure :: Changed OutputStructure}
+  | OutputsDontMatch { outputStructure :: Changed OutputStructure}
+  | DerivationDiff
+      { outputStructure :: Changed OutputStructure
+      , outputsDiff     :: OutputsDiff
+      , platformDiff    :: Maybe (Changed Platform)
+        -- ^ Will be Nothing, if Platform does not change
+      , builderDiff     :: Maybe (Changed Builder)
+        -- ^ Will be Nothing, if Builder does not change
+      , argumentsDiff   :: Maybe (TextDiff NonEmpty)
+        -- ^ Will be Nothing, if arguments are equal
+      , sourcesDiff     :: SourcesDiff
+      , inputsDiff      :: InputsDiff
+      , envDiff         :: Maybe EnvironmentDiff
+        -- ^ Will be Nothing, if environment comparison is skipped
+      }
+  deriving stock Generic
+  deriving anyclass ToJSON
```

Comment:
I think it might be useful to also provide `FromJSON` instances so that `nix-diff` can also be used as a library and these diffs can be de-serialized back to Haskell types since raw JSON representation is quite hardly-readable